README
This is the POWERPC ArchC 2.22functional model.

This model has the system call emulation functions implemented,
so it is a good idea to turn on the ABI option.

To use acsim, the interpreted simulator:

    acsim powerpc.ac -abi               (create the simulator)
    make -f Makefile.archc              (compile)
    powerpc.x --load=<file-path> [args] (run an application)

The [args] are optional arguments for the application.

There are two formats recognized for application <file-path>:
- ELF binary matching ArchC specifications
- hexadecimal text file for ArchC


To generate binary utilities use:

    acbingen.sh -i<abs-install-path> -a<arch-name> powerpc.ac

This will generate the tools source files using the architeture
name <arch-name> (if omitted, powerpc is used), copy them to the
binutils source tree, build and install them into the directory
<abs-install-path> (which -must- be an absolute path).
Use "acbingen.sh -h" to get information about the command-line
options available.


For more information visit http://www.archc.org


CHANGELOG:
==========

Version 1.0.1:

- Updated syscall emulation to work with new version of libac_sysc
  (this new version of libac_sysc is used by GCC 4.8 cross compiler and above)

Version 1.0.0:

- ArchC 2.2 compliant

Version 0.7.4:

- Fixed gdb_funcs bug when using ArchC 2.0 or later
- ArchC 2.1 compliant

Version 0.7.3:

- Added binary utilities support files
- Adjusted mullhw syntax
- ArchC 2.0 compliant


Version 0.7.2-archc2.0beta3:

- Added license headers


Version 0.7.1-archc2.0beta1

- Model compliant with ArchC 2.0beta1


Version 0.7.1:

- Included optional properties to optimize compiled simulation
- Changed dprintf function (conflict with native dprintf())
- Changed SC1 type (to include the LEV field)
- Changed XL2 type (to include the BH field)
- Added assembly language information


Version 0.7.0:

- Model passed selected Mediabench and Mibench applications
- Included file for GDB integration


HISTORY.md
## 2.4.0

* Revision numbers following the ArchC release
* Instructions with cycles annotations
* Two new .ac files to use with MPSoCBench (block and nonblock)
* powerpc_isa.cpp using the reserved work DATA_PORT to data request. See the [commit message](https://github.com/ArchC/powerpc/commit/18be8f932ee67c61795a331aa4682c866c864358).
* Interrupt handler support. It is inactive in standalone simulator.

[Full changelog](https://github.com/ArchC/powerpc/compare/v2.3.0...v2.4.0)

## 2.3.0
* Revision numbers following the ArchC release
+ Added id register for core identification

[Full changelog](https://github.com/ArchC/powerpc/compare/v1.0.2...v2.3.0)

## 1.0.1
* Updated syscall emulation to work with new version of libac_sysc
  (this new version of libac_sysc is used by GCC 4.8 cross compiler and above)

[Full changelog](https://github.com/ArchC/powerpc/compare/v1.0.0...v1.0.1)

## 1.0.0
* ArchC 2.2 compliant

[Full changelog](https://github.com/ArchC/powerpc/compare/v0.7.4...v1.0.0)

## 0.7.4
* Fixed gdb_funcs bug when using ArchC 2.0 or later
* ArchC 2.1 compliant

[Full changelog](https://github.com/ArchC/powerpc/compare/v0.7.3...v0.7.4)

## 0.7.3
+ Added binary utilities support files
* Adjusted mullhw syntax
* ArchC 2.0 compliant

[Full changelog](https://github.com/ArchC/powerpc/compare/v0.7.2...v0.7.3)

## 0.7.2-archc2.0beta3

+ Added license headers

[Full changelog](https://github.com/ArchC/powerpc/compare/v0.7.1-1...v0.7.2)

## 0.7.1-archc2.0beta1

* Model compliant with ArchC 2.0beta1

## 0.7.1

+ Included optional properties to optimize compiled simulation
* Changed dprintf function (conflict with native dprintf())
* Changed SC1 type (to include the LEV field)
* Changed XL2 type (to include the BH field)
+ Added assembly language information

## 0.7.0

* Model passed selected Mediabench and Mibench applications
+ Included file for GDB integration


powerpc_gdb_funcs.cpp
/**
 * @file      powerpc_gdb_funcs.cpp
 * @author    Bruno Corsi dos Santos
 *
 *            The ArchC Team
 *            http://www.archc.org/
 *
 *            Computer Systems Laboratory (LSC)
 *            IC-UNICAMP
 *            http://www.lsc.ic.unicamp.br
 *
 * @version   1.0
 * @date      Mon, 19 Jun 2006 15:50:48 -0300
 * 
 * @brief     The ArchC POWERPC functional model.
 * 
 * @attention Copyright (C) 2002-2006 --- The ArchC Team
 *
 */

#include "powerpc.H"

using namespace powerpc_parms;

int powerpc::nRegs(void) {
  return 104;
}


ac_word powerpc::reg_read( int reg ) {
  unsigned int n;

  if ( ( reg >= 0 ) && ( reg < 32 ) )
    return GPR.read( reg );
  else {
    switch (reg) {
    
      case 96:
	n=ac_pc;
      break;

      /* case ??:
        n=MSR.read();
	break; */

      case 98:
        n=CR.read();
      break;

      case 99:
        n=LR.read();
      break;

      case 100:
        n=CTR.read();
      break;

      case 101:
        n=XER.read();
      break;

      default:
        return 0;
      break;

    }
    
    return(n);
    
  }
}


void powerpc::reg_write( int reg, ac_word value ) {

  /* general purpose registers */
  if ( ( reg >= 0 ) && ( reg < 32 ) )
    GPR.write(reg,value);
  else {
    switch (reg) {

    case 96:
      ac_pc=value;
    break;

    /*    case ??:
      MSR.write(value);
    break;
    */

    case 98:
      CR.write(value);
    break;

    case 99:
      LR.write(value);
    break;

    case 100:
      CTR.write(value);
    break;

    case 101:
      XER.write(value);
    break;

    default:
      /* No completely implemented register */
      break;

    }
  }
}


unsigned char powerpc::mem_read( unsigned int address ) {
  return IM->read_byte( address );
}


void powerpc::mem_write( unsigned int address, unsigned char byte ) {
  IM->write_byte( address, byte );
}



powerpc_syscall.cpp
/**
 * @file      powerpc_syscall.cpp
 * @author    Bruno Corsi dos Santos
 *
 *            The ArchC Team
 *            http://www.archc.org/
 *
 *            Computer Systems Laboratory (LSC)
 *            IC-UNICAMP
 *            http://www.lsc.ic.unicamp.br
 *
 * @version   1.0
 * @date      Mon, 19 Jun 2006 15:50:48 -0300
 * 
 * @brief     The ArchC POWERPC functional model.
 * 
 * @attention Copyright (C) 2002-2006 --- The ArchC Team
 * 
 * This program is free software; you can redistribute it and/or modify 
 * it under the terms of the GNU General Public License as published by 
 * the Free Software Foundation; either version 2 of the License, or 
 * (at your option) any later version. 
 * 
 * This program is distributed in the hope that it will be useful, 
 * but WITHOUT ANY WARRANTY; without even the implied warranty of 
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
 * GNU General Public License for more details. 
 * 
 * You should have received a copy of the GNU General Public License 
 * along with this program; if not, write to the Free Software 
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#include "powerpc_syscall.H"

using namespace powerpc_parms;

void powerpc_syscall::get_buffer(int argn, unsigned char* buf, unsigned int size)
{
  unsigned int addr = GPR.read(3+argn); 

  for (unsigned int i = 0; i<size; i++, addr++) {
    buf[i] = DATA_PORT->read_byte(addr);
  }
}

void powerpc_syscall::set_buffer(int argn, unsigned char* buf, unsigned int size)
{
  unsigned int addr = GPR.read(3+argn);

  for (unsigned int i = 0; i<size; i++, addr++) {
    DATA_PORT->write_byte(addr, buf[i]);
  }
}

void powerpc_syscall::set_buffer_noinvert(int argn, unsigned char* buf, unsigned int size)
{
  unsigned int addr = GPR.read(3+argn);

  for (unsigned int i = 0; i<size; i+=4, addr+=4) {
    DATA_PORT->write(addr, *(unsigned int *) &buf[i]);
  }
}

int powerpc_syscall::get_int(int argn)
{
  return GPR.read(3+argn);
}

void powerpc_syscall::set_int(int argn, int val)
{
  GPR.write(3+argn, val);
}

void powerpc_syscall::return_from_syscall()
{
  unsigned int oldr1;
  unsigned int oldr31;
//  oldr1=MEM.read(GPR.read(1));
//  oldr31=MEM.read(GPR.read(1)+28);
//  GPR.write(1,oldr1);
//  GPR.write(31,oldr31);
  ac_pc=LR.read();
}

void powerpc_syscall::set_prog_args(int argc, char **argv)
{
  int i, j, base;

  unsigned int ac_argv[30];
  char ac_argstr[512];

  base = AC_RAM_END - 512;
  for (i=0, j=0; i<argc; i++) {
    int len = strlen(argv[i]) + 1;
    ac_argv[i] = base + j;
    memcpy(&ac_argstr[j], argv[i], len);
    j += len;
  }

  //Write argument string
  GPR.write(3, AC_RAM_END-512);
  set_buffer(0, (unsigned char*) ac_argstr, 512);

  //Write string pointers
  GPR.write(3, AC_RAM_END-512-120);
  set_buffer_noinvert(0, (unsigned char*) ac_argv, 120);

  //Set r3 to the argument count
  GPR.write(3, argc);

  //Set r4 to the string pointers
  GPR.write(4, AC_RAM_END-512-120);
}



